src/cpu/instr.rs

```diff
@@ -40,30 +94,230 @@ struct Operation {
+            execute: |enc, cpu| {
+                let _out:u32 = cpu.state.reg[enc.rs1 as usize] as u32 >> enc.imm as u32;
+                cpu.state.reg[enc.rd as usize] = _out as u16;
+                cpu.state.flags = gen_flag(false, cpu.state.reg[enc.rs1 as usize] as u32, cpu.state.reg[enc.rs1 as usize] as u32, _out);
+                cpu.state.pc = cpu.state.pc+1; 
+            },
+            repr: |enc| format!("sri r{0}, r{1}, {2}", enc.rd, enc.rs1, enc.imm),
+        });
+        m.insert(Opcode::DIV as u8, Operation {
+           execute: |enc, cpu| {
+               cpu.state.reg[enc.rd as usize] = cpu.state.reg[enc.rs1 as usize] / cpu.state.reg[enc.rs2 as usize];
+               cpu.state.pc = cpu.state.pc+1;
+           },
+           repr: |enc| format!("div r{0}, r{1}, r{2}", enc.rd, enc.rs1, enc.rs2),
+        });
+        m.insert(Opcode::MUL as u8, Operation {
+            execute: |enc, cpu| {
+                cpu.state.reg[enc.rd as usize] = cpu.state.reg[enc.rs1 as usize] * cpu.state.reg[enc.rs2 as usize];
+                cpu.state.pc = cpu.state.pc+1;
+            },
+            repr: |enc| format!("mul r{0}, r{1}, r{2}", enc.rd, enc.rs1, enc.rs2),
+        });
+        m.insert(Opcode::MOD as u8, Operation {
+            execute: |enc, cpu| {
+                cpu.state.reg[enc.rd as usize] = cpu.state.reg[enc.rs1 as usize] % cpu.state.reg[enc.rs2 as usize];
+                cpu.state.pc = cpu.state.pc+1;
+            },
+            repr: |enc| format!("mod r{0}, r{1}, r{2}", enc.rd, enc.rs1, enc.rs2),
+        });
+        m.insert(Opcode::CMP as u8, Operation {
+            execute: |enc, cpu| {
+                let _out: u32 = cpu.state.reg[enc.rs1 as usize] as u32 - cpu.state.reg[enc.rs2 as usize] as u32;
+                cpu.state.flags = gen_flag(true, cpu.state.reg[enc.rs1 as usize] as u32, cpu.state.reg[enc.rs1 as usize] as u32, _out)
+            },
+            repr: |enc| format!("cmp r{}, r{}", enc.rs1, enc.rs2),
+        });
+        m.insert(Opcode::CMI as u8, Operation {
+            execute: |enc, cpu| {
+                let _out: u32 = cpu.state.reg[enc.rs1 as usize] as u32 - enc.imm as u32;
+                cpu.state.flags = gen_flag(true, cpu.state.reg[enc.rs1 as usize] as u32, enc.imm as u32, _out);
+            },
+            repr: |enc| format!("cmp r{}, {}", enc.rs1, enc.imm),
+        });
+        m.insert(Opcode::JAL as u8, Operation {
+            execute: |enc, cpu| {
+                cpu.state.reg[enc.rd as usize] = cpu.state.pc;
+                cpu.state.pc = enc.imm; 
+            },
+            repr: |enc| format!("jal {}", enc.imm), //probably should be better
+        });
+        m.insert(Opcode::JMP as u8, Operation {
+            execute: |enc, cpu| {
+                let jmp_code: u8 = ((enc.rs1 << 3) as u8) +  enc.rd as u8;
+                let cpu_flags: Flags = convert_to_bitflags(cpu.state.flags);
+
+                match jmp_code {
```

Comment:
Consider returning Boolean value from that match, that means whether the cpu should jump and handle it in one place later.

PC should be incremented by one in opposite case.

Reply:
You can make match return a value. `let jump_conition_met = match ...`. (`0x0 => true, 0x1 => cpu_flags.contains(Flag.C)`). No need for those all if statements